test/jasmine/tests/splom_test.js

```diff
@@ -785,30 +907,48 @@ describe('Test splom interactions:', function() {
             }, {
                 values: ['lyndon', 'richard', 'gerald', 'jimmy', 'ronald', 'george', 'bill', 'georgeW', 'barack', 'donald']
             }, {
-                values: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+                values: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
+                axis: {type: 'category'}
+            }, {
+                values: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
+                axis: {type: 'log'}
             }]
         }];
 
         Plotly.plot(gd, data).then(function() {
-            expect(gd.layout.xaxis.type).toBe('linear');
-            expect(gd.layout.xaxis2.type).toBe('category');
-            expect(gd.layout.xaxis3).toBeUndefined();
-            expect(gd.layout.yaxis.type).toBe('category');
-            expect(gd.layout.yaxis2.type).toBe('linear');
-            expect(gd.layout.yaxis3).toBeUndefined();
+            _assertAxisTypes('no upper half / no diagonal', {
+                xaxes: ['linear', 'category', undefined, null],
+                fullXaxes: ['linear', 'category', 'category', null],
+                yaxes: [null, 'category', undefined, undefined],
+                fullYaxes: [null, 'category', 'category', 'log']
+            });
 
             return Plotly.restyle(gd, {
                 'showupperhalf': true,
                 'diagonal.visible': true
             });
         })
         .then(function() {
-            expect(gd.layout.xaxis.type).toBe('linear');
-            expect(gd.layout.xaxis2.type).toBe('category');
-            expect(gd.layout.xaxis3.type).toBe('linear');
-            expect(gd.layout.yaxis.type).toBe('linear');
-            expect(gd.layout.yaxis2.type).toBe('category');
-            expect(gd.layout.yaxis3.type).toBe('linear');
+            _assertAxisTypes('full grid', {
+                xaxes: ['linear', 'category', undefined, undefined],
+                fullXaxes: ['linear', 'category', 'category', 'log'],
+                yaxes: ['linear', 'category', undefined, undefined],
+                fullYaxes: ['linear', 'category', 'category', 'log']
+            });
+
+            return Plotly.restyle(gd, {
+                'showlowerhalf': false,
+                'diagonal.visible': false
+            });
+        })
+        .then(function() {
+            // TODO should we delete layout.xaxis and layout.yaxis4 here?
```

Comment:
no, we should leave them there in case the user brings diagonal or lowerhalf back later